drivers/timer/nrf_rtc_timer.c

```diff
@@ -117,21 +118,80 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
 	ticks = MAX(MIN(ticks - 1, (s32_t)MAX_TICKS), 0);
 
 	k_spinlock_key_t key = k_spin_lock(&lock);
-	u32_t cyc, t = counter();
+	u32_t cyc, dt, t = counter();
+	bool flagged = false;
 
 	/* Round up to next tick boundary */
-	cyc = ticks * CYC_PER_TICK + counter_sub(t, last_count);
+	cyc = ticks * CYC_PER_TICK + 1 + counter_sub(t, last_count);
 	cyc += (CYC_PER_TICK - 1);
 	cyc = (cyc / CYC_PER_TICK) * CYC_PER_TICK;
 	cyc += last_count;
 
-	if (counter_sub(cyc, t) < MIN_DELAY) {
-		cyc += CYC_PER_TICK;
+	/* Per NRF docs, the RTC is guaranteed to trigger a compare
+	 * event if the comparator value to be set is at least two
+	 * cycles later than the current value of the counter.  So if
+	 * we're three or more cycles out, we can set it blindly.  If
+	 * not, check the time again immediately after setting: it's
+	 * possible we "just missed it" and can flag an immediate
+	 * interrupt.  Or it could be exactly two cycles out, which
+	 * will have worked.  Otherwise, there's no way to get an
+	 * interrupt at the right time and we have to slip the event
+	 * by one clock cycle (or we could spin, but this is a slow
+	 * clock and spinning for a whole cycle can be thousands of
+	 * instructions!)
+	 *
+	 * You might ask: why not set the comparator first and then
+	 * check the timer synchronously to see if we missed it, which
+	 * would avoid the need for a slipped cycle.  That doesn't
+	 * work, the states overlap inside the counter hardware.  It's
+	 * possible to set a comparator value of "N", issue a DSB
+	 * instruction to flush the pipeline, and then immediately
+	 * read a counter value of "N-1" (i.e. the comparator is still
+	 * in the future), and yet still not receive an interrupt at
+	 * least on nRF52.  Some experimentation on nrf52840 shows
+	 * that you need to be early by about 400 processor cycles
+	 * (about 1/5th of a RTC cycle) in order to reliably get the
+	 * interrupt.  The docs say two cycles, they mean two cycles.
+	 */
```

Comment:
I would propose following algorithm. It would be common for ZLI and nonZLI case:
```
if (counter_sub(cyc, t) > 1) {
	set_comparator(cyc);
	t = counter();
   /* decrement to cover cyc == t which will return 0xffffff */
	dt = counter_sub(cyc - 1, t);
	if (dt > 0x7fffff) {
		/* EXPIRED, trigger irq */
	} else if (dt == 0) {
       /* cc is set to expire in one tick from now. it is possible that 
        * it was set too late thus move it 1 tick further */
		if (tick_delayed_alarm(c, t)) {
			/* EXPIRED, trigger irq */
		}
	} else {
		/* set on time, RTC interrupt will come */
	}
} else { /* now + 1 case, delay by one tick */
	if (tick_delayed_alarm(c, t)) {
		/* EXPIRED, trigger irq */
	}
}
```
and additional function for setting cc for value delayed by one tick:
```
static bool tick_delayed_alarm(u32_t c, u32_t now)
{
	set_comparator(c + 1);
	if (now != counter()) {
		/* 1 tick already elapsed */
		return true;
	}
	/* set on time, RTC interrupt will come */
	return false;
}
```

Reply:
This doesn't actually work.  That was the first trick I tried, actually.  The hardware states overlap.

It's possible to read a counter value of "N", set a comparator value of "N+1", immediately issue a DSB instruction to be sure the pipeline is flushed, and then read the counter back to see a "N".  You would think that the write got there in time, but it didn't and the interrupt won't arrive.

You have to be about 20% of a RTC cycle **ahead** of the transition to get the interrupt to trigger reliably, which the documentation pads to a full RTC cycle (i.e. you have to set N+2 to make it work).  That is what the current PR implements.

Reply:
> It's possible to read a counter value of "N", set a comparator value of "N+1", immediately issue a DSB instruction to be sure the pipeline is flushed, and then read the counter back to see a "N". You would think that the write got there in time, but it didn't and the interrupt won't arrive.

True, but here we would be setting `now+2` since function is called when we see that cc `now+1`. In that case change to `n+2` and attempt to write. If after writing `now` has changed that means that 1 tick elapsed and we can say that requested timeout elapsed. Of course that is if 1 tick timeout means next tick and not `n+2` tick.

I've been testing the scenario where current CC has `now+1` value and we reconfigure to something in the future before CC value is reached. If we do that more than ~15us (half of 32k tick?) before the deadline then we won't get that interrupt, if later then old CC value will generate interrupt. So it seemed like CC is somehow latched between the ticks.